README.md

```diff
@@ -31,57 +35,118 @@ To get started,
 
-The gem includes the following objects:
+## Usage
+
+*IMPORTANT NOTE:* You must create your own `MessageProcessor` class
+for this gem to work in your application.
+
+See more information below.
+
+### `MessageQueue::Producer`
+
+This is a class
+which provides an adapter to the
+fake and real NSQ producers.
+These are used to
+write messages onto the queue:
+
+```ruby
+message_data = {
+  "event_type": "heartbeat",
+  "data": {
+    "key": "value"
+  }
+}
 
-### [`MessageQueue`]
+producer = MessageQueue::Producer.new(
+  nsqd: ENV.fetch('NSQD_TCP_ADDRESS'),
+  topic: topic,
+).connection
 
-This is an adapter class
-which takes a required `topic` string
-and provides entry points
-for the queue's message producer and consumer classes.
+producer.write(message_data.to_json)
+```
 
-The queue strategy used
+The mock/real strategy used
 can be switched
 by adding an environment variable
 to your application:
 
 ```ruby
-if ENV['FAKE_QUEUE'] == true
-  FakeMessageQueue
-else
-  NsqMessageQueue
-end
+# for the fake
+ENV['FAKE_QUEUE'] == true
+
+# for the real thing, comment it out or delete it or make it anything but true
+# ENV['FAKE_QUEUE'] == false
 ```
 
-[`MessageQueue`]: lib/fastly_nsq/message_queue.rb
+### `MessageQueue::Consumer`
+This is a class
+which provides an adapter to the
+fake and real NSQ consumers.
+These are used to
+read messages off of the queue:
 
+```ruby
+consumer = MessageQueue::Consumer.new(
+  topic: 'topic',
+  channel: 'channel'
+).connection
+
+message = consumer.size #=> 1
+message = consumer.pop
+message.body #=>'hey this is my message!'
+message.finish
+message = consumer.size #=> 0
+```
 
-### [`NsqMessageQueue`]
+As above,
+the mock/real strategy used
+can be switched by adding
+the `FAKE_QUEUE` environment variable.
 
-This strategy
-creates a connection
-to `nsq-ruby`'s
-`Nsq::Producer` and `Nsq::Consumer` classes.
+### `MessageQueue::Listener`
 
-[`NsqMessageQueue`]: lib/fastly_nsq/nsq_message_queue.rb
+To process the next message on the queue:
 
+```ruby
+topic = 'user_created'
+channel = 'my_consuming_service'
 
-### [`FakeMessageQueue`]
+MessageQueue::Listener.new(topic: topic, channel: channel).process_next_message
+```
 
-This strategy
-mocks the connection
-to NSQ for testing purposes.
+This will pop the next message
+off of the queue
+and send it to `MessageProcessor.new(message).start`.
 
-It adheres to the same API
-as `NsqMessageQueue`.
+To initiate a blocking loop to process messages continuously:
 
-[`FakeMessageQueue`]: lib/fastly_nsq/fake_message_queue.rb
+```ruby
+topic = 'user_created'
+channel = 'my_consuming_service'
 
+MessageQueue::Listener.new(topic: topic, channel: channel).start
+```
 
-*IMPORTANT NOTE:* You must create your own `MessageProcessor` class
-for this gem to work in your application.
+This will block until
+there is a new message on the queue,
+      pop the next message
+      off of the queue
+      and send it to `MessageProcessor.new(message).start`.
```

Comment:
considering it is expected of the `start` message to `finish` the message, perhaps it should go by another name. `process`?